drivers/hmis/spec/requests/hmis/project_spec.rb

```diff
@@ -123,6 +126,76 @@
       expect(result.dig('data', 'project', 'access')).to include('canDeleteProject' => true)
     end
   end
+
+  describe 'project assessments query' do
+    before(:each) do
+      hmis_login(user)
+    end
+
+    let(:project_assessments_query) do
+      <<~GRAPHQL
+        query GetProjectAssessments(
+          $id: ID!
+          $limit: Int = 10
+          $offset: Int = 0
+        ) {
+          project(id: $id) {
+            assessments(
+              limit: $limit
+              offset: $offset
+            ) {
+              offset
+              limit
+              nodesCount
+              nodes {
+                id
+                inProgress
+                role
+                definition {
+                  title
+                }
+                enrollment {
+                  entryDate
+                  exitDate
+                }
+              }
+            }
+          }
+        }
+      GRAPHQL
+    end
+
+    it 'resolves both WIP and non-WIP assessments on the project' do
+      response, result = post_graphql(id: p1.id) { project_assessments_query }
+      expect(response.status).to eq 200
+      records = result.dig('data', 'project', 'assessments', 'nodes')
+      expect(records.size).to eq(2)
+      expect(records[0]['inProgress']).to be_falsy
+      expect(records[1]['inProgress']).to be_truthy
+    end
+
+    it 'resolves assessments in a reasonable amount of time' do
+      100.times.map do
+        c = create :hmis_hud_client_complete, data_source: ds1, user: u1
+        e = create :hmis_hud_enrollment, data_source: ds1, project: p1, client: c, user: u1
+        5.times.map do
+          create :hmis_wip_custom_assessment, enrollment: e, client: c
+          create :hmis_custom_assessment, enrollment: e, client: c
+        end
+      end
+
+      expect do
+        _response, result = post_graphql(id: p1.id) { project_assessments_query }
+        expect(result.dig('data', 'project', 'assessments', 'nodesCount')).to eq 1002
+      end.to perform_under(500).ms
```

Comment:
100 records is a lot, I don't think we need that many in order to prevent an N+1 situation from arising. A  `make_database_queries` assertion is probably more likely to help us identify that if it comes up (for example if we are resolving 50 assessments vs 10 assessments the query count shouldn't be all that different).

It looks like we likely _do_ have an N+1 issue, though. This code will do a normal db lookup (without using a graphql loader) https://github.com/greenriver/hmis-warehouse/blob/5dd4c19cbb4978efcd39635ed6d26f3c5f89a3a8/drivers/hmis/app/graphql/types/hmis_schema/assessment.rb#L72

That's out of scope to address here, but relates to the convo around Custom Assessments being linked to Definitions. For now you could either
a. pass `definition: ` to the assessment factories. then they should be loaded with the graphql loaders, since the association exists. add a test for db queries that shows there is no n+1
b. add a test with a high db count and note about why

Reply:
happy to chat through that if you'd like!

Reply:
Hm, yeah maybe it would be helpful to chat as I'm not sure I have 100% understood.
- It looks like passing `definition:` to the assessment factories doesn't change the number of DB queries being made. I added a test I knew would fail to get it to print out all the queries it's making, here is the list:
```SELECT "users".* FROM "users" WHERE "users"."deleted_at" IS NULL AND "users"."id" = $1 ORDER BY "users"."id" ASC LIMIT $2
UPDATE "users" SET "last_activity_at" = $1 WHERE "users"."id" = $2
SELECT "data_sources".* FROM "data_sources" WHERE "data_sources"."deleted_at" IS NULL AND "data_sources"."hmis" IS NOT NULL AND "data_sources"."hmis" = $1 LIMIT $2
SELECT "hmis_roles"."id" FROM "hmis_roles" WHERE "hmis_roles"."deleted_at" IS NULL AND "hmis_roles"."can_view_project" = TRUE
SELECT "hmis_access_controls"."access_group_id" FROM "hmis_access_controls" INNER JOIN "hmis_user_groups" ON "hmis_access_controls"."user_group_id" = "hmis_user_groups"."id" INNER JOIN "hmis_user_group_members" ON "hmis_user_groups"."id" = "hmis_user_group_members"."user_group_id" WHERE "hmis_access_controls"."deleted_at" IS NULL AND "hmis_user_groups"."deleted_at" IS NULL AND "hmis_user_group_members"."deleted_at" IS NULL AND "hmis_user_group_members"."user_id" = $1 AND "hmis_access_controls"."role_id" = $2
SELECT "Project"."id" FROM "Project" WHERE "Project"."DateDeleted" IS NULL AND "Project"."id" IN (SELECT "hmis_group_viewable_entities"."entity_id" FROM "hmis_group_viewable_entities" WHERE "hmis_group_viewable_entities"."deleted_at" IS NULL AND "hmis_group_viewable_entities"."collection_id" = $1 AND "hmis_group_viewable_entities"."entity_type" = $2)
SELECT "Project"."id" FROM "Organization" INNER JOIN "Project" ON "Project"."DateDeleted" IS NULL AND "Project"."data_source_id" = "Organization"."data_source_id" AND "Project"."OrganizationID" = "Organization"."OrganizationID" WHERE "Organization"."DateDeleted" IS NULL AND "Organization"."id" IN (SELECT "hmis_group_viewable_entities"."entity_id" FROM "hmis_group_viewable_entities" WHERE "hmis_group_viewable_entities"."deleted_at" IS NULL AND "hmis_group_viewable_entities"."collection_id" = $1 AND "hmis_group_viewable_entities"."entity_type" = $2)
SELECT "Project"."id" FROM "data_sources" INNER JOIN "Project" ON "Project"."DateDeleted" IS NULL AND "Project"."data_source_id" = "data_sources"."id" WHERE "data_sources"."deleted_at" IS NULL AND "data_sources"."id" IN (SELECT "hmis_group_viewable_entities"."entity_id" FROM "hmis_group_viewable_entities" WHERE "hmis_group_viewable_entities"."deleted_at" IS NULL AND "hmis_group_viewable_entities"."collection_id" = $1 AND "hmis_group_viewable_entities"."entity_type" = $2)
SELECT "Project".* FROM "Project" WHERE "Project"."DateDeleted" IS NULL AND "Project"."id" = $1 AND "Project"."data_source_id" = $2 AND "Project"."id" = $3 LIMIT $4
SELECT "hmis_roles".* FROM "hmis_roles" INNER JOIN "hmis_access_controls" ON "hmis_roles"."id" = "hmis_access_controls"."role_id" INNER JOIN "hmis_user_groups" ON "hmis_access_controls"."user_group_id" = "hmis_user_groups"."id" INNER JOIN "hmis_user_group_members" ON "hmis_user_groups"."id" = "hmis_user_group_members"."user_group_id" WHERE "hmis_roles"."deleted_at" IS NULL AND "hmis_access_controls"."deleted_at" IS NULL AND "hmis_user_groups"."deleted_at" IS NULL AND "hmis_user_group_members"."deleted_at" IS NULL AND "hmis_user_group_members"."user_id" = $1
SELECT "Project"."id", "hmis_group_viewable_entities"."collection_id" FROM "Project" INNER JOIN "hmis_group_viewable_entity_projects" ON "hmis_group_viewable_entity_projects"."project_id" = "Project"."id" INNER JOIN "hmis_group_viewable_entities" ON "hmis_group_viewable_entities"."deleted_at" IS NULL AND "hmis_group_viewable_entities"."id" = "hmis_group_viewable_entity_projects"."group_viewable_entity_id" WHERE "Project"."DateDeleted" IS NULL AND "Project"."id" = $1
SELECT hmis_roles.*, hmis_access_controls.access_group_id AS access_group_id FROM "hmis_roles" INNER JOIN "hmis_access_controls" ON "hmis_roles"."id" = "hmis_access_controls"."role_id" INNER JOIN "hmis_user_groups" ON "hmis_access_controls"."user_group_id" = "hmis_user_groups"."id" INNER JOIN "hmis_user_group_members" ON "hmis_user_groups"."id" = "hmis_user_group_members"."user_group_id" INNER JOIN "hmis_access_controls" "access_controls_hmis_roles" ON "access_controls_hmis_roles"."deleted_at" IS NULL AND "access_controls_hmis_roles"."role_id" = "hmis_roles"."id" WHERE "hmis_roles"."deleted_at" IS NULL AND "hmis_access_controls"."deleted_at" IS NULL AND "hmis_user_groups"."deleted_at" IS NULL AND "hmis_user_group_members"."deleted_at" IS NULL AND "hmis_user_group_members"."user_id" = $1
SELECT "hmis_roles"."id" FROM "hmis_roles" WHERE "hmis_roles"."deleted_at" IS NULL AND "hmis_roles"."can_view_enrollment_details" = $1 AND "hmis_roles"."can_view_project" = $2
SELECT "Project"."id", "Project"."ProjectID" FROM "Project" WHERE "Project"."DateDeleted" IS NULL AND "Project"."id" = $1 AND "Project"."data_source_id" = $2
SELECT COUNT(*) FROM "CustomAssessments" INNER JOIN "Enrollment" ON "Enrollment"."DateDeleted" IS NULL AND "Enrollment"."EnrollmentID" = "CustomAssessments"."EnrollmentID" AND "Enrollment"."PersonalID" = "CustomAssessments"."PersonalID" AND "Enrollment"."data_source_id" = "CustomAssessments"."data_source_id" LEFT OUTER JOIN "hmis_wips" ON "hmis_wips"."deleted_at" IS NULL AND "hmis_wips"."source_id" = "Enrollment"."id" AND "hmis_wips"."source_type" = $1 LEFT OUTER JOIN "hmis_wips" "wips_Enrollment" ON "wips_Enrollment"."deleted_at" IS NULL AND "wips_Enrollment"."source_id" = "Enrollment"."id" AND "wips_Enrollment"."source_type" = $2 WHERE "CustomAssessments"."DateDeleted" IS NULL AND ("Enrollment"."ProjectID" = '200' OR "hmis_wips"."project_id" = 908936) AND "Enrollment"."DateDeleted" IS NULL AND ("hmis_wips"."project_id" IN (908936) OR "Enrollment"."ProjectID" IN ('200'))
SELECT "CustomAssessments".* FROM "CustomAssessments" INNER JOIN "Enrollment" ON "Enrollment"."DateDeleted" IS NULL AND "Enrollment"."EnrollmentID" = "CustomAssessments"."EnrollmentID" AND "Enrollment"."PersonalID" = "CustomAssessments"."PersonalID" AND "Enrollment"."data_source_id" = "CustomAssessments"."data_source_id" LEFT OUTER JOIN "hmis_wips" ON "hmis_wips"."deleted_at" IS NULL AND "hmis_wips"."source_id" = "Enrollment"."id" AND "hmis_wips"."source_type" = $1 LEFT OUTER JOIN "hmis_wips" "wips_Enrollment" ON "wips_Enrollment"."deleted_at" IS NULL AND "wips_Enrollment"."source_id" = "Enrollment"."id" AND "wips_Enrollment"."source_type" = $2 WHERE "CustomAssessments"."DateDeleted" IS NULL AND ("Enrollment"."ProjectID" = '200' OR "hmis_wips"."project_id" = 908936) AND "Enrollment"."DateDeleted" IS NULL AND ("hmis_wips"."project_id" IN (908936) OR "Enrollment"."ProjectID" IN ('200')) LIMIT $3 OFFSET $4
SELECT "hmis_client_projects".* FROM "hmis_client_projects" WHERE ("hmis_client_projects"."data_source_id" = 64090 AND "hmis_client_projects"."EnrollmentID" = '500' OR "hmis_client_projects"."data_source_id" = 64090 AND "hmis_client_projects"."EnrollmentID" = '501')
SELECT "Project".* FROM "Project" WHERE "Project"."DateDeleted" IS NULL AND "Project"."id" = $1
SELECT "Enrollment".* FROM "Enrollment" WHERE "Enrollment"."DateDeleted" IS NULL AND ("Enrollment"."EnrollmentID" = '500' AND "Enrollment"."PersonalID" = '100' AND "Enrollment"."data_source_id" = 64090 OR "Enrollment"."EnrollmentID" = '501' AND "Enrollment"."PersonalID" = '101' AND "Enrollment"."data_source_id" = 64090)
SELECT "Project".* FROM "Project" WHERE "Project"."DateDeleted" IS NULL AND "Project"."data_source_id" = 64090 AND "Project"."ProjectID" = '200'
SELECT "hmis_form_processors".* FROM "hmis_form_processors" WHERE "hmis_form_processors"."custom_assessment_id" IN ($1, $2, $3, $4)
SELECT "hmis_wips".* FROM "hmis_wips" WHERE "hmis_wips"."deleted_at" IS NULL AND "hmis_wips"."source_type" = $1 AND "hmis_wips"."source_id" = $2
SELECT "hmis_form_definitions".* FROM "hmis_form_definitions" WHERE "hmis_form_definitions"."deleted_at" IS NULL AND "hmis_form_definitions"."id" = $1
SELECT "Exit".* FROM "Exit" WHERE "Exit"."DateDeleted" IS NULL AND ("Exit"."EnrollmentID" = '500' AND "Exit"."PersonalID" = '100' AND "Exit"."data_source_id" = 64090 OR "Exit"."EnrollmentID" = '501' AND "Exit"."PersonalID" = '101' AND "Exit"."data_source_id" = 64090)
INSERT INTO "hmis_activity_logs" ("user_id","data_source_id","ip_address","session_hash","variables","referer","operation_name","header_page_path","header_client_id","header_enrollment_id","header_project_id","created_at","resolved_fields","resolved_at") VALUES (4562, 64090, '127.0.0.1', '806328772d4133a4b3254298cb49d9ba', '{"id":908936}', NULL, NULL, NULL, NULL, NULL, NULL, '2024-04-09 16:11:45.381751', '{"Project/908936":[],"Assessment/6857":[],"Assessment/6858":[],"Assessment/6859":[],"Assessment/6860":[],"Enrollment/260714":[],"Enrollment/260715":[]}', '2024-04-09 16:11:45.723424') RETURNING "id"
```
- It looks like the line you indicated in `assessment.rb` isn't reached, because a form definition is returned by `load_ar_association` on the previous line. And I _think_ that is true whether or not we pass `definition:` to the assessment factory
- Should I remove the test for # of db queries from `drivers/hmis/spec/models/hmis/hud/project_spec.rb` entirely? Or keep it there as well as adding one here?
- Should the test in this file that tests `perform_under` instantiate lots of records, or just a couple (the normal ones instantiated at the top of this file)?